src/file/SDL_iostream.c

```diff
@@ -558,6 +558,160 @@ SDL_IOStream *SDL_IOFromFP(FILE *fp, bool autoclose)
 }
 #endif // !HAVE_STDIO_H && !defined(SDL_PLATFORM_WINDOWS)
 
+#if !defined(SDL_PLATFORM_WINDOWS)
+
+// Functions to read/write file descriptors. Not used for windows.
+
+#ifdef HAVE_LSEEK64
+#define lseek_off_t off64_t
+#define lseek       lseek64
+#else
+#define lseek_off_t off_t
+#endif
+
+typedef struct IOStreamFDData
+{
+    int fd;
+    bool autoclose;
+    bool regular_file;
+} IOStreamFDData;
+
+static Sint64 SDLCALL fd_seek(void *userdata, Sint64 offset, SDL_IOWhence whence)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+    int fdwhence;
+
+    switch (whence) {
+    case SDL_IO_SEEK_SET:
+        fdwhence = SEEK_SET;
+        break;
+    case SDL_IO_SEEK_CUR:
+        fdwhence = SEEK_CUR;
+        break;
+    case SDL_IO_SEEK_END:
+        fdwhence = SEEK_END;
+        break;
+    default:
+        SDL_SetError("Unknown value for 'whence'");
+        return -1;
+    }
+
+	lseek_off_t result = lseek(iodata->fd, (lseek_off_t)offset, fdwhence);
+	if (result < 0) {
+		SDL_SetError("Couldn't get stream offset: %s", strerror(errno));
+	}
+	return result;
+}
+
+static size_t SDLCALL fd_read(void *userdata, void *ptr, size_t size, SDL_IOStatus *status)
+{
+    IOStreamFDData *iodata = (IOStreamFDData *) userdata;
+    ssize_t bytes = read(iodata->fd, ptr, size);
```

Comment:
Strictly speaking this should be something like

```suggestion
    ssize_t bytes;
    do {
        bytes = read(iodata->fd, ptr, size);
    } while (bytes < 0 && errno == EINTR);
```

because `read()` is an example of a syscall that can be interrupted by a signal if we are unlucky.

Reply:
glibc has a macro for this, `TEMP_FAILURE_RETRY`, but unfortunately that's non-standard and non-portable.